File: 2020/Day22/crabCombat_part2.R
#!/usr/bin/Rscript

#clear workspace
rm(list = ls())

# Libraries

#------------------------------------------------------------------------------------------
#Load Data
#------------------------------------------------------------------------------------------

#Use this if you're running from the command line:
# mydata <- strsplit(readLines(file("stdin"))," ")

#Use this if you're opening this repo as a R project, using relative paths:
# mydata <- readLines('2020/Day22/input_test1.txt')
mydata <- readLines('2020/Day22/input_test1.txt')

#------------------------------------------------------------------------------------------
#The Game
#------------------------------------------------------------------------------------------

#decks
player_idx <- which(grepl("Player", mydata, fixed = TRUE)==TRUE)

player1 <-as.numeric(mydata[(player_idx[1]+1):(player_idx[2]-2)])
player2 <-as.numeric(mydata[(player_idx[2]+1):(length(mydata))])

#rounds of the game
player1_wins <- function(player1,player2){
  prev_rounds <- list(player1, player2)
  while(length(player1)>0 && length(player2)>0){
    value_1 <- player1[1]
    value_2 <- player2[1]
    
    if(length(player1) >= value_1 && length(player2) >= value_2){
      
      result <- player1_wins(player1[1:value_1],player2[1:value_2])
      player1 <- result[[2]]
      player2 <- result[[3]]
      if (result[[1]]){
        player1 <- append(player1,value_1) 
        player1 <- append(player1,value_2) 
      }else{
        player2 <- append(player2,value_2) 
        player2 <- append(player2,value_1)
      }
      
    }else{
      if(value_1 > value_2){
        player1 <- append(player1,value_1) 
        player1 <- append(player1,value_2)
      }else{
        player2 <- append(player2,value_2) 
        player2 <- append(player2,value_1)
      }
    }
    
    if(length(player1)!=0 && length(player2)!=0){
      return(list(TRUE,player1,player2))
    }else{
      prev_rounds <- append(prev_rounds,list(player1, player2))
    }
    
    
  }
  
  if(length(player1)>length(player2)){
    return(list(TRUE,player1,player2))
  }else{return(list(FALSE,player1,player2))}
  
}


# score
calculate_score <- function(playerx){
  count = 0
  for(i in 1:length(playerx)){
    count = count+ playerx[i]*(length(playerx)-i)
  }
  return(count)
}



player1_wins(player1,player2)
calculate_score(player1_wins(player1,player2)[[2]]) #player 1
calculate_score(player1_wins(player1,player2)[[3]]) #player 2

File: 2020/Day22/crabCombat_part2.py
def p1_wins(p1, p2, part2):
    prev_rounds = [(p1.copy(), p2.copy())]

    while len(p1) > 0 and len(p2) > 0:
        val1 = p1.pop(0)
        val2 = p2.pop(0)
        if part2 and len(p1) >= val1 and len(p2) >= val2:
            if p1_wins(p1[0:val1].copy(), p2[0:val2].copy(), True):
                p1.append(val1)
                p1.append(val2)
            else:
                p2.append(val2)
                p2.append(val1)
        else:
            if val1 > val2:
                p1.append(val1)
                p1.append(val2)
            else:
                p2.append(val2)
                p2.append(val1)
        if part2 and (p1, p2) in prev_rounds:
            return True
        elif part2:
            prev_rounds.append((p1.copy(), p2.copy()))

    if len(p1) > len(p2):
        return True
    else:
        return False


def score(p):
    count = 0
    for i, val in enumerate(p):
        count += val * (len(p) - i)
    return count


with open("input.txt") as f:
    lines = f.readlines()

p1 = []
for i in range(25):
    p1.append(int(lines[i+1]))

p2 = []
for i in range(25):
    p2.append(int(lines[i+28]))

p1_copy = p1.copy()
p2_copy = p2.copy()

if p1_wins(p1, p2, False):
    print("Player 1 won first game with score:", score(p1))
else:
    print("Player 2 won first game with score:", score(p2))

if p1_wins(p1_copy, p2_copy, True):
    print("Player 1 won second game with score:", score(p1_copy))
else:
    print("Player 2 won second game with score:", score(p2_copy))
